### DISCUSSION TITLE
Feature Request: Calculation of Second- and Third-Order Nonlinear Optical Properties

### DISCUSSION THREAD

**Discussion by dyzheng** [deepmodeling/abacus-develop#5288] [General]
### Background

From https://github.com/abacusmodeling/abacus-develop/issues/298

### Describe the solution you'd like

.

### Task list only for developers

- [ ] Notice possible changes of behavior
- [ ] Explain the changes of codes in core modules of ESolver, HSolver, ElecState, Hamilt, Operator or Psi

### Task list for Issue attackers (only for developers)

- [ ] Review and understand the proposed feature and its importance.
- [ ] Research on the existing solutions and relevant research articles/resources.
- [ ] Discuss with the team to evaluate the feasibility of implementing the feature.
- [ ] Create a design document outlining the proposed solution and implementation details.
- [ ] Get feedback from the team on the design document.
- [ ] Develop the feature following the agreed design.
- [ ] Write unit tests and integration tests for the feature.
- [ ] Update the documentation to include the new feature.
- [ ] Perform code review and address any issues.
- [ ] Merge the feature into the main branch.
- [ ] Monitor for any issues or bugs reported by users after the feature is released.
- [ ] Address any issues or bugs reported by users and continuously improve the feature.

**Comment by WHUweiqingzhou** [comment]
@jingan-181 could PYATB calculate these properties now?

**Comment by jingan-181** [comment]
1. The nonlinear optical properties currently supported by PYATB are mainly second-order types, including shift current, injection current (or circular photovoltaic effect), second harmonic generation (SHG), Berry curvature dipole, and Pockels effect. Third-order optical properties have not been covered yet.

2. The functions requested in the https://github.com/abacusmodeling/abacus-develop/issues/298, third harmonic generation (THG), two-photon absorption (TPA), Stimulated Raman Scattering (SRS), and Kerr effects, are not currently available in PYATB.
